src/blog/internet-computer-web-app-decentralized-database-architecture.md

```diff
@@ -0,0 +1,345 @@
+
+*****
+
+### The “Nowadays”
+
+Before digging into the Internet Computer, let's review first a common concept, a good old' school fashion way approach.
+
+Nowadays, server-less or not, the data persistence is often solve with one database for all information.
+
+![](https://cdn-images-1.medium.com/max/1600/1*eWuHlblTCJMdrJVNllri9Q.png)
+
+Users use the (web) application. It calls endpoint(s) to persist and read data and, it returns the results to the users.
+
+In a more fresh microservice architecture style,  the application(s) is split as a collection of services but, to some extension, the outcome is still based on previous concept.
+
+![](https://cdn-images-1.medium.com/max/1600/1*3mAmSM4_l4QDF0aK2I_esA.png)
+
+Users use the (web) application. It calls a dedicated service which call endpoint(s) to persist and read data and, returns the results to the users.
+
+Regardless of these architectures, mono- or micro-services, the persistence of the data finds place for all the users at the same place.
+
+*****
+
+#### Code Snippet
+
+I do not know the exact architecture of Google [Firestore](https://firebase.google.com/docs/firestore/quickstart) but, from a user and outside point of view perspective, it looks like a mono-database hosted in the cloud.
+
+If we develop, for example, an application that list the kind of animals our users own, we would most probably define a db collection of `pets` to collect their entries.
+
+![](https://cdn-images-1.medium.com/max/1600/1*I4CP0N8GvViWCuAzRLa3UA.png)
+
+To query and persist the data, we would then implement a getter (`get` ) and setter (`add` ) functions in our web application that would call the cloud database.
+
+```javascript
+import firebase from 'firebase/app';
+import 'firebase/firestore';
+
+const firestore = firebase.firestore();
+
+const get = async (entryId) => {
+  const snapshot = 
+            await firestore.collection('pets').doc(entryId).get();
+
+  if (snapshot.exists) {
+    console.log(snapshot.data());
+  }
+};
+
+const add = async (userId, dog, cat) => {
+  const {id: entryId} = await firestore.collection('pets').add({
+    userId,
+    dog,
+    cat
+  });
+
+  console.log(entryId);
+};
+```
+
+All users would use the same entry point and, their data would also be saved in the same collection (`firestore.collection('pets')` ).
+
+*****
+
+### The “Conservative”
+
+The Internet Computer does not provide a built-in out of the box database solution, that is not the goal of the DFINITY foundation (my understanding). They provide a futuristic blockchain network, that can run web applications too.
+
+It is up to the makers to build features on the top. They can implement features and, unleash these on the IC through canister smart contracts.
+
+I picture my self a "canister" has a container that run a set of functions in the cloud, except that it is actually a decentralized blockchain network 😉. It can basically do anything that can be implemented, from computing to data persistence.
```

Comment:
```suggestion
I picture a "canister" as a container that runs a set of functions in the cloud, except that it is actually a decentralized blockchain network 😉. It can basically do anything that can be implemented, from computing to data persistence.
```